datumaro/cli/util/project.py

```diff
@@ -22,17 +28,104 @@ def generate_next_file_name(basename, basedir='.', sep='.', ext=''):
-    matches = [match for match in (pattern.match(n) for n in names) if match]
+_dataset_revpath_regex = None
+_full_revpath_regex = None
+def parse_full_revpath(s: str, ctx_project: Optional[Project]):
+    if ctx_project:
+        env = ctx_project.env
+    else:
+        env = Environment()
+
+    def parse_dataset_pathspec(s: str):
+        global _dataset_revpath_regex
+        if not _dataset_revpath_regex:
+            _dataset_revpath_regex = re.compile(
+                "%s(:%s)?" % \
+                (
+                    r"(?P<dataset_path>[^:]+)",
+                    r"(?P<format>.+)"
+                )
+            )
+
+        match = re.fullmatch(_dataset_revpath_regex, s)
+        if not match:
+            raise ValueError("Failed to recognize dataset pathspec in '%s'" % s)
+        match = { k: unescape(v, escapes=escapes) if v else v
+            for k, v in match.groupdict().items() }
+
+        path = match["dataset_path"]
+        format = match["format"]
+        return Dataset.import_from(path, format, env=env)
+
+    def parse_revspec(s: str):
+        global _full_revpath_regex
+        if not _full_revpath_regex:
+            _full_revpath_regex = re.compile(
+                "(%(proj_path)s(@%(rev)s)?(:%(source)s)?)" % \
+                {
+                    'proj_path': r"(?P<proj_path>[^@:]+)",
+                    'rev': r"(?P<rev>[^:]+)",
+                    'source': r"(?P<source>.+)",
+                }
+            )
+
+        match = re.fullmatch(_full_revpath_regex, s)
+        if not match:
+            raise ValueError("Failed to recognize revspec in '%s'" % s)
+        match = { k: unescape(v, escapes=escapes) if v else v
+            for k, v in match.groupdict().items() }
+
+        proj_path = match["proj_path"]
+        rev = match["rev"]
+        source = match["source"]
+
+        assert proj_path
+        if rev:
+            project = load_project(proj_path)
 
-    max_idx = max([cast(match[1], int, 0) for match in matches], default=None)
-    if max_idx is None:
-        if default is not None:
-            idx = sep + str(default)
+            # proj_path is either proj_path or rev or source name
+        elif Project.find_project_dir(proj_path):
+            project = load_project(proj_path)
+        elif ctx_project:
+            project = ctx_project
+            if project.is_ref(proj_path):
+                rev = proj_path
+            elif not source:
+                source = proj_path
         else:
-            idx = ''
+            raise ProjectNotFoundError("Failed to find project at '%s'. " \
+                "Specify project path with '-p/--project' or in the "
+                "target pathspec." % proj_path)
+
+        tree = project.get_rev(rev)
+        return tree.make_dataset(source)
+
+
+    # Escape colons used in absolute paths on Windows
+    escapes = [(':\\', r'%%driveroot_bs%%')]
+    escapes = [(':/', r'%%driveroot_s%%')]
```

Comment:
This could create problems if one of the strings on the right is present in the original path. And if the intent is to allow colon-slash sequences to be a part of the path, a simpler way to do so would be to incorporate that directly into the regex: replace `[^:]+` with `(?: [^:] | :[/\\] )+` (assuming `re.VERBOSE`).

Reply:
> This could create problems if one of the strings on the right is present in the original path.

Maybe, but such input is incorrect.

I implemented the solution you propose and refactored the code a bit. But I fear it might return to the initial version even on little changes - on addition of subset selector, for example. Hopefully, it won't.

Reply:
> Maybe, but such input is incorrect.

Why would it be incorrect? These can be valid paths.